src/org/zaproxy/zap/extension/websocket/manualsend/WebSocketReopenSend.java

```diff
@@ -0,0 +1,388 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.zaproxy.zap.extension.websocket.manualsend;
+
+import org.apache.commons.httpclient.URI;
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.db.DatabaseException;
+import org.parosproxy.paros.extension.history.ExtensionHistory;
+import org.parosproxy.paros.extension.manualrequest.MessageSender;
+import org.parosproxy.paros.model.HistoryReference;
+import org.parosproxy.paros.model.Model;
+import org.parosproxy.paros.model.Session;
+import org.parosproxy.paros.network.HttpMalformedHeaderException;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpSender;
+import org.parosproxy.paros.view.View;
+import org.zaproxy.zap.PersistentConnectionListener;
+import org.zaproxy.zap.ZapGetMethod;
+import org.zaproxy.zap.extension.httppanel.Message;
+import org.zaproxy.zap.extension.websocket.WebSocketObserver;
+import org.zaproxy.zap.extension.websocket.WebSocketProxy;
+import org.zaproxy.zap.extension.websocket.WebSocketProxyV13;
+import org.zaproxy.zap.extension.websocket.WebSocketMessage;
+import org.zaproxy.zap.model.SessionStructure;
+import org.zaproxy.zap.network.HttpRedirectionValidator;
+import org.zaproxy.zap.network.HttpRequestConfig;
+import org.zaproxy.zap.utils.Pair;
+
+import javax.net.ssl.SSLException;
+
+import java.awt.EventQueue;
+import java.io.IOException;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Stack;
+
+/**
+ * Reopen a closed connection. Sending the same handshake as the closed connection is going to be re-connected.
+ * External listener is set at {@link WebSocketReopenSend#addPersistentConnectionListener(PersistentConnectionListener)}
+ * is able to hear and establish new WebSocket connection by HTTP response.
+ */
+public class WebSocketReopenSend implements MessageSender, WebSocketObserver{
+    private static final Logger LOGGER = Logger.getLogger(WebSocketReopenSend.class);
+
+    private HttpSender delegate;
+    private boolean useTrackingSessionState;
+    private boolean followRedirects;
+    private List<PersistentConnectionListener> persistentConnectionListeners;
+    private List<WebSocketReopenListener> webSocketReopenListeners;
+    private HashMap<String, Stack<Integer>> reopenedPool;
+    public final static String SEC_WEBSOCKET_KEY = "Sec-WebSocket-Key";
```

Comment:
This could be moved to `WebSocketMessage` instead and renamed to `HTTP_HEADER_SEC_WEBSOCKET_KEY`.